streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamPrinter.java

```diff
@@ -0,0 +1,126 @@
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.kstream.internals;
+
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.streams.processor.AbstractProcessor;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+
+import java.io.PrintStream;
+
+
+class KStreamPrinter<K, V> implements ProcessorSupplier<K, V> {
+
+    private final PrintStream printStream;
+    private Serde<?> keySerde;
+    private Serde<?> valueSerde;
+    private static final String NULL_KEY = "null_key";
+    private boolean notStandardOut;
+
+
+    KStreamPrinter(PrintStream printStream, Serde<?> keySerde, Serde<?> valueSerde) {
+        this.keySerde = keySerde;
+        this.valueSerde = valueSerde;
+        if (printStream == null) {
+            this.printStream = System.out;
+        } else {
+            this.printStream = printStream;
+            notStandardOut = true;
+        }
+    }
+
+    KStreamPrinter(PrintStream printStream) {
+        this(printStream, null, null);
+    }
+
+    KStreamPrinter(Serde<?> keySerde, Serde<?> valueSerde) {
+        this(null, keySerde, valueSerde);
+    }
+
+    KStreamPrinter() {
+        this(null, null, null);
+    }
+
+    @Override
+    public Processor<K, V> get() {
+        return new KStreamPrinterProcessor(this.printStream, this.keySerde, this.valueSerde);
+    }
+
+
+    private class KStreamPrinterProcessor extends AbstractProcessor<K, V> {
+        private final PrintStream printStream;
+        private Serde<?> keySerde;
+        private Serde<?> valueSerde;
+        private ProcessorContext processorContext;
+
+        private KStreamPrinterProcessor(PrintStream printStream, Serde<?> keySerde, Serde<?> valueSerde) {
+            this.printStream = printStream;
+            this.keySerde = keySerde;
+            this.valueSerde = valueSerde;
+        }
+
+        @Override
+        public void init(ProcessorContext context) {
+            this.processorContext = context;
+
+            if (this.keySerde == null) {
+                keySerde = this.processorContext.keySerde();
+            }
+
+            if (this.valueSerde == null) {
+                valueSerde = this.processorContext.valueSerde();
+            }
+        }
+
+        @Override
+        public void process(K key, V value) {
+            Object keyToPrint = maybeDeserialize(key, keySerde.deserializer());
+            Object valueToPrint = maybeDeserialize(value, valueSerde.deserializer());
+
```

Comment:
These two lines will be changed to expect the correct types.  If not a meaningful message will be provided (not just a ClassCastException)